generator/internal/parser/routing_info.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parser
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/googleapis/google-cloud-rust/generator/internal/api"
+	"google.golang.org/genproto/googleapis/api/annotations"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+func parseRoutingAnnotations(methodID string, m *descriptorpb.MethodDescriptorProto) ([]*api.RoutingInfo, error) {
+	var info []*api.RoutingInfo
+	extensionId := annotations.E_Routing
+	if !proto.HasExtension(m.GetOptions(), extensionId) {
+		return info, nil
+	}
+
+	rule := proto.GetExtension(m.GetOptions(), extensionId).(*annotations.RoutingRule)
+	var errs []error
+	for _, routing := range rule.GetRoutingParameters() {
+		new, err := parseRoutingInfo(methodID, routing)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		info = append(info, new)
+	}
+	if len(errs) != 0 {
+		return nil, errors.Join(errs...)
+	}
+	return info, nil
+}
+
+func parseRoutingInfo(methodID string, routing *annotations.RoutingParameter) (*api.RoutingInfo, error) {
+	pathTemplate := routing.GetPathTemplate()
+	fieldName := routing.GetField()
+	info, err := parseRoutingPathTemplate(fieldName, pathTemplate)
+	if err != nil {
+		return nil, fmt.Errorf("%w, method=%s", err, methodID)
+	}
+	return info, nil
+}
+
+func parseRoutingPathTemplate(fieldName, pathTemplate string) (*api.RoutingInfo, error) {
+	fieldPath := strings.Split(fieldName, ".")
+	if pathTemplate == "" {
+		if len(fieldPath) != 1 {
+			return nil, fmt.Errorf("fieldName (%q) cannot have '.' when path_template is empty", fieldName)
+		}
+		info := &api.RoutingInfo{
+			FieldPath: fieldPath,
+			Name:      fieldName,
+			Matching: api.RoutingPathSpec{
+				Segments: []string{api.RoutingSegmentMulti},
+			},
+		}
+		return info, nil
+	}
+	pos := 0
+	prefix, width := parseRoutingPrefix(pathTemplate[pos:])
+	pos += width
+	if !strings.HasPrefix(pathTemplate[pos:], "{") {
+		return nil, fmt.Errorf("expected '{', found=%s", pathTemplate[pos:])
+	}
+	pos += 1
+	name, match, width, err := parseRoutingVariable(fieldName, pathTemplate[pos:])
+	if err != nil {
+		return nil, err
+	}
+	pos += width
+	if !strings.HasPrefix(pathTemplate[pos:], "}") {
+		return nil, fmt.Errorf("expected '}', found=%s", pathTemplate[pos:])
+	}
+	pos += 1
+	suffix := api.RoutingPathSpec{}
+	if strings.HasPrefix(pathTemplate[pos:], "/") {
```

Comment:
comment: I originally did not understand what was wrong with this test case

https://github.com/googleapis/google-cloud-rust/blob/aba4513e1d1f37cf5b4584c2ca401570fd5f7864/generator/internal/parser/routing_info_test.go#L334

... but I see in the spec that segments are either literal, or variable. It cannot mix the two.

so :+1: